File: Javascript/Arrays/Problem15.js

let arr=[2,6,5,8,11];


let target=14;
// 1 . Brute Force solution



for(let i=0;i<arr.length;i++){
  for(let j=i+1;j<arr.length;j++){
    if(arr[i]+arr[j]==target){
      console.log(i,j);
      break;
    }
  }
}



// 2 Better solution

let map=new Map()


for(let i=0;i<arr.length;i++){
  
  let sub=target-arr[i];
  
  if(map.has(sub)){
    let val=map.get(sub)
    console.log(i,val)
  }
  else{
    map.set(arr[i],i)
  }
}


// 3 Optimal solution but in case of YES and NO






arr.sort((a,b)=>a-b)





let i=0;
let j=arr.length-1;

while(i<j){
  
console.log(i,j);
  if(arr[i]+arr[j]==target){
 
    console.log(i,j);
    break;
  }
  else if(arr[i]+arr[j]<target){
    i++;
  }
  else{
    j--;
  }
}

File: C++/Arrays/Problem15.cpp


// 1. Brute Force

#include <iostream>
using namespace std;

int main() 
{
   int n=6;
   int arr[n]={2,6,5,8,11};
   int target=14;
   
   for(int i=0;i<n;i++){
     for(int j=i+1;j<n;j++){
       
       if(arr[i]+arr[j]==target){
         
         std::cout << i<<" "<< j << std::endl;
         
         break;
         
       }
     }
   }
   
   return 0;
}



// 2. Better Solution

#include <iostream>
#include <bits/stdc++.h>
using namespace std;

int main() 
{
   int n=6;
   int arr[n]={2,6,5,8,11};
   int target=14;
   
   unordered_map<int, int> mpp;
    for (int i = 0; i < n; i++) {
        int num = arr[i];
        int moreNeeded = target - num;
        if (mpp.find(moreNeeded) != mpp.end()) {
            return {mpp[moreNeeded], i};
        }
        mpp[num] = i;
    }
    
    
  
   
   return 0;
}

